### ISSUE TITLE
false redundant_closure_call for "poor man's try"

### ISSUE THREAD

**Issue by CAD97** [rust-lang-nursery/rust-clippy#3421]
```rust
use std::num::ParseIntError;

#[derive(Debug)]
struct SillyError {
    cause: ParseIntError,
}

fn silly(a: &str, b: &str) -> Result<(u32, u32), SillyError> {
    match (|| {
        Ok((a.parse()?, b.parse()?))
    })() {
        Ok(this) => Ok(this),
        Err(cause) => Err(SillyError { cause }),
    }
}
```

[playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2015&gist=a45b68ac04832b4568dfcc0e351eaf2d)

```
warning: Try not to call a closure in the expression where it is declared.
  --> src/main.rs:9:11
   |
9  |       match (|| {
   |  ___________^
10 | |         Ok((a.parse()?, b.parse()?))
11 | |     })() {
   | |________^
   |
   = note: #[warn(clippy::redundant_closure_call)] on by default
   = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#redundant_closure_call
help: Try doing something like: 
   |
9  |     match {
10 |         Ok((a.parse()?, b.parse()?))
11 |     } {
   |
```

which doesn't work, as expected:

```
error[E0271]: type mismatch resolving `<u32 as std::str::FromStr>::Err == SillyError`
  --> src/main.rs:10:15
   |
10 |         Ok((a.parse()?, b.parse()?))
   |               ^^^^^ expected struct `std::num::ParseIntError`, found struct `SillyError`
   |
   = note: expected type `std::num::ParseIntError`
              found type `SillyError`

error[E0271]: type mismatch resolving `<u32 as std::str::FromStr>::Err == SillyError`
  --> src/main.rs:10:27
   |
10 |         Ok((a.parse()?, b.parse()?))
   |                           ^^^^^ expected struct `std::num::ParseIntError`, found struct `SillyError`
   |
   = note: expected type `std::num::ParseIntError`
              found type `SillyError`
```

(As it turns out, it was indeed redundant in the case that triggered this, as I didn't need to prevent the error from bubbling, and its desired type was identical to the function's type. But this is still a false positive for a real pattern.)

**Comment by mikerite** [comment]
Duplicate of #1684
